File: javascript/clock/clock.js
const leftPad = require('left-pad');

const minutesPerHour = 60;
const hoursPerDay = 24;
const minutesPerDay = minutesPerHour * hoursPerDay;
const maxMinutes = minutesPerDay;

const mod = function mod(x, d) {
  let rem = x % d;
  if (rem < 0) {
    rem += d;
  }
  return rem;
};

const calcMinutes = function calcMinutes(hours, minutes) {
  const realMinutes = minutes || 0;
  return (hours * minutesPerHour) + realMinutes;
};

const getMinutes = function getMinutes(totalMinutes) {
  return totalMinutes % minutesPerHour;
};

const getHours = function getHours(totalMinutes) {
  return Math.floor(totalMinutes / minutesPerHour);
};

const clockToString = function clockToString(totalMinutes) {
  const minutes = getMinutes(totalMinutes);
  const hours = getHours(totalMinutes);
  const minutesStr = leftPad(minutes, 2, '0');
  const hoursStr = leftPad(hours, 2, '0');
  return `${hoursStr}:${minutesStr}`;
};

const Clock = {
  at: function at(hours, minutes) {
    const totalMinutes = mod(calcMinutes(hours, minutes), maxMinutes);

    this._totalMinutes = totalMinutes;

    this.toString = function toString() {
      return clockToString(totalMinutes);
    };

    this.equals = function equals(other) {
      /* If we try to use this._totalMinutes === other._totalMinutes, we
         run into problems with `this` ... */
      return totalMinutes === other._totalMinutes;
    };

    this.plus = function plus(diffMinutes) {
      return at(0, totalMinutes + diffMinutes);
    };

    this.minus = function minus(diffMinutes) {
      return at(0, totalMinutes - diffMinutes);
    };

    return this;
  },

};

module.exports = Clock;

File: csharp/clock/Clock.cs
using System;

public struct Clock
{
    private const int MINUTES_PER_HOUR = 60;
    private const int MAX_HOURS = 24;
    private const int MAX_MINUTES = MINUTES_PER_HOUR * MAX_HOURS;

    private int totalMinutes;

    public Clock(int minutes)
    {
        totalMinutes = RollOver(minutes);
    }

    public Clock(int hours, int minutes)
    {
        totalMinutes = RollOver(hours * MINUTES_PER_HOUR + minutes);
    }

    public int Hours
    {
        get
        {
            return totalMinutes / MINUTES_PER_HOUR;
        }
    }

    public int Minutes
    {
        get
        {
            return totalMinutes % MINUTES_PER_HOUR;
        }
    }

    public int TotalMinutes
    {
        get
        {
            return totalMinutes;
        }
    }

    public Clock Add(int minutesToAdd)
    {
        return new Clock(TotalMinutes + minutesToAdd);
    }

    public Clock Subtract(int minutesToSubtract)
    {
        return new Clock(TotalMinutes + minutesToSubtract);
    }

    private static int RollOver(int minutes)
    {
        return Mod(minutes, MAX_MINUTES);
    }

    private static int Mod(int number, int divisor)
    {
        int x = number % divisor;
        if (x < 0)
        {
            x += divisor;
        }
        return x;
    }

    public override string ToString()
    {
        return $"{Hours:d2}:{Minutes:d2}";
    }
}